### PR TITLE
Fix panic in match checking

### PR DESCRIPTION
**Pull request by jonas-schievink**
Fixes https://github.com/rust-analyzer/rust-analyzer/issues/4416

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Looks like there's another panic that this doesn't fix:

```rust
pub enum Category {
    Infinity,
    Zero,
}

fn panic(a: Category, b: Category) {
    match (a, b) {
        (Category::Infinity, Category::Infinity) | (Category::Zero, Category::Zero) => {}

        (Category::Infinity | Category::Zero, _) => {}
    }
}
```

Fixing that one causes a spurious "missing match arm" diagnostic that I'm not sure how to fix.

File: crates/ra_hir_ty/src/_match.rs [hunk 1]
```diff
@@ -2036,16 +2061,18 @@ mod false_negatives {
...
-        check_no_diagnostic(content);
+        ",
+        );
     }
 
     #[test]
     fn expr_loop_missing_arm() {
-        let content = r"
+        // This is a false negative.
+        // We currently infer the type of `loop { break Foo::A }` to `!`, which
```
Comment (flodiebold) [inline crates/ra_hir_ty/src/_match.rs:2071]: we do support break-with-value now, the problem in this test is that `Foo` doesn't exist.

**Comment by flodiebold** [comment]
I don't have the match checking algorithm in cache enough to see how that other example would crash or give a spurious error :sweat_smile:

**Pull request merged**